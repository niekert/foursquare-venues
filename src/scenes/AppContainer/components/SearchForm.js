import React from 'react';
import { NormalCard } from 'style/Cards';
import { IconButton } from 'style/Buttons';
import { MyLocationIcon } from 'shared/components/icons';
import styled from 'styled-components';
import { prop } from 'styled-tools';

const Wrapper = styled.div`
  position: absolute;
  z-index: 10;
  min-width: 350px;
  left: ${prop('theme.spacing.2')};
  top: ${prop('theme.spacing.2')};
`;

const UseLocationButton = IconButton.extend`
  color: '#333';
`;

const SearchContainer = styled.div`
  position: relative;
  display: flex;
`;

const SearchInput = styled.input.attrs({
  type: 'text',
})`
  height: 32px;
  width: 100%;
  border: none;
  font-size: 16px;
  border: none;
  outline: none;
`;

function SearchForm() {
  return (
    <Wrapper>
      <NormalCard>
        <SearchContainer>
          <SearchInput placeholder="Find venues in..." />
          <UseLocationButton>
            <MyLocationIcon />
          </UseLocationButton>
        </SearchContainer>
      </NormalCard>
    </Wrapper>
  );
}

export default SearchForm;
