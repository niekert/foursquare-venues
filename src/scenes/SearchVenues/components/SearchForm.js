import React, { Component } from 'react';
import { func, oneOf } from 'prop-types';
import { NormalCard } from 'style/Cards';
import { smallUp } from 'style/breakpoints';
import VenuesList from 'scenes/VenuesList';
import styled from 'styled-components';
import { SEARCH_TYPE_QUERY, SEARCH_TYPE_CURRENT_GEO } from 'app-constants';
import { prop } from 'styled-tools';
import UseLocationButton from './UseLocationButton';

const Wrapper = styled.div`
  position: absolute;
  z-index: 10;
  min-width: 350px;
  padding: 0 ${prop('theme.spacing.1')};
  left: 0;
  right: 0;
  top: ${prop('theme.spacing.2')};

  ${smallUp`
    left: ${prop('theme.spacing.2')};
    right: initial;
    padding: 0;
  `};
`;

const SearchContainerWrapper = styled.div`
  position: relative;
  display: flex;
  justify-content: space-between;
`;

const StyledForm = styled.form`
  flex: 1;
`;

const SearchInput = styled.input.attrs({
  type: 'text',
})`
  height: 32px;
  width: 100%;
  border: none;
  font-size: 16px;
  border: none;
  outline: none;
`;

class SearchForm extends Component {
  static propTypes = {
    fetchVenuesBySearch: func.isRequired,
    fetchVenuesByGeoLocation: func.isRequired,
    activeSearchType: oneOf([SEARCH_TYPE_QUERY, SEARCH_TYPE_CURRENT_GEO]),
  };

  static defaultProps = {
    activeSearchType: null,
  };

  state = {
    isGeoLocationLoading: false,
    search: '',
  };

  onSearchChange = e => {
    this.setState({
      search: e.target.value,
    });
  };

  getGeoLocation = e => {
    e.stopPropagation();
    e.preventDefault();

    this.setState({ isGeoLocationLoading: true });

    navigator.geolocation.getCurrentPosition(position => {
      this.setState({
        isGeoLocationLoading: false,
      });

      this.props.fetchVenuesByGeoLocation(position);
    });
  };

  submitForm = e => {
    e.preventDefault();

    this.props.fetchVenuesBySearch(this.state.search);
  };

  render() {
    const { activeSearchType } = this.props;
    const { search, isGeoLocationLoading } = this.state;

    return (
      <Wrapper>
        <NormalCard>
          <SearchContainerWrapper>
            <StyledForm onSubmit={this.submitForm}>
              <SearchInput
                placeholder="Find venues near..."
                onChange={this.onSearchChange}
                value={search}
              />
            </StyledForm>
            <UseLocationButton
              isLoading={isGeoLocationLoading}
              isActive={activeSearchType === SEARCH_TYPE_CURRENT_GEO}
              onClick={this.getGeoLocation}
            />
          </SearchContainerWrapper>
          <VenuesList />
        </NormalCard>
      </Wrapper>
    );
  }
}

export default SearchForm;
