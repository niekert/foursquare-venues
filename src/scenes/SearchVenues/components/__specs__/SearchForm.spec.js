import React from 'react';
import { shallow, mount } from 'enzyme';
import SearchForm from '../SearchForm';

jest.mock('scenes/VenuesList', () => 'VenuesList');

describe('SearchForm', () => {
  let props;
  beforeEach(() => {
    props = {
      fetchVenuesBySearch: jest.fn(),
      fetchVenuesByGeoLocation: jest.fn(),
      activeSearchType: null,
    };
  });

  test('renders correctly', () => {
    const wrapper = shallow(<SearchForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('fetches venues by search when submitting form', () => {
    const wrapper = mount(<SearchForm {...props} />);
    const mockSearchQuery = 'mock_search_query';

    wrapper
      .find('input')
      .simulate('change', { target: { value: mockSearchQuery } });

    wrapper.find('form').simulate('submit');

    expect(props.fetchVenuesBySearch).toHaveBeenCalledTimes(1);
    expect(props.fetchVenuesBySearch).toHaveBeenLastCalledWith(mockSearchQuery);
  });
});
