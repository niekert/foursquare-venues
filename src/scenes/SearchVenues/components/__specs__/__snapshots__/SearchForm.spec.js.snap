// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders correctly 1`] = `
<SearchForm__Wrapper>
  <Cards__NormalCard>
    <SearchForm__SearchContainerWrapper>
      <SearchForm__StyledForm
        onSubmit={[Function]}
      >
        <SearchForm__SearchInput
          onChange={[Function]}
          placeholder="Find venues near..."
          value=""
        />
      </SearchForm__StyledForm>
      <branch(UseLocationButton)
        isActive={false}
        isLoading={false}
        onClick={[Function]}
      />
    </SearchForm__SearchContainerWrapper>
    <VenuesList />
  </Cards__NormalCard>
</SearchForm__Wrapper>
`;
